packages/react-spinbutton/src/components/SpinButton/useSpinButton.tsx

```diff
@@ -0,0 +1,316 @@
+
+  React.useEffect(() => {
+    let newTextValue;
+    if (value !== undefined) {
+      const roundedValue = precisionRound(value, precision);
+      newTextValue = displayValue ?? String(roundedValue);
+      internalState.current.value = roundedValue;
+      setAtBound(getBound(roundedValue, min, max));
+    } else {
+      newTextValue = String(precisionRound(currentValue, precision));
+      internalState.current.value = currentValue;
+    }
+    setTextValue(newTextValue);
+  }, [value, displayValue, currentValue, precision, setAtBound, min, max]);
+
+  const handleWheel = (e: React.WheelEvent<HTMLDivElement>) => {
+    if (props.disabled) {
+      return;
+    }
+
+    const spinDir = e.deltaY > 0 ? 'down' : 'up';
+    // TODO: Might want to debounce this
+    setSpinState(spinDir);
+    stepValue(e, spinDir);
+
+    setWheelTimeout(() => {
+      setSpinState('rest');
+    }, 250);
+  };
+
+  const handleInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+    if (inputType === 'spinners-only') {
+      return;
+    }
+
+    if (!internalState.current.previousTextValue) {
+      internalState.current.previousTextValue = textValue;
+    }
+
+    const newValue = e.target.value;
+    setTextValue(newValue);
+  };
+
+  const stepValue = (e: SpinButtonChangeEvent, direction: 'up' | 'down' | 'upPage' | 'downPage') => {
+    const dir = direction === 'up' || direction === 'upPage' ? 1 : -1;
+    const stepSize = direction === 'upPage' || direction === 'downPage' ? stepPage : step;
+    const val = internalState.current.value;
+
+    let newValue = val + stepSize * dir;
+    if (!Number.isNaN(newValue)) {
+      newValue = clampWhenInRange(val, newValue, min, max);
+    }
+
+    commit(e, newValue);
+
+    if (internalState.current.spinState !== 'rest') {
+      setStepTimeout(() => {
+        // Ease the step speed a bit
+        internalState.current.spinTime += internalState.current.spinDelay;
+        internalState.current.spinDelay = lerp(
+          DEFAULT_SPIN_DELAY_MS,
+          MIN_SPIN_DELAY_MS,
+          internalState.current.spinTime / MAX_SPIN_TIME_MS,
+        );
+        stepValue(e, direction);
+      }, internalState.current.spinDelay);
+    }
+  };
+
+  const handleIncrementMouseDown = (e: React.MouseEvent<HTMLButtonElement>) => {
+    internalState.current.spinState = 'up';
+    stepValue(e, 'up');
+  };
+
+  const handleDecrementMouseDown = (e: React.MouseEvent<HTMLButtonElement>) => {
+    internalState.current.spinState = 'down';
+    stepValue(e, 'down');
+  };
+
+  const handleStepMouseUpOrLeave = (e: React.MouseEvent<HTMLButtonElement>) => {
+    clearStepTimeout();
+    internalState.current.spinState = 'rest';
+    internalState.current.spinDelay = DEFAULT_SPIN_DELAY_MS;
+    internalState.current.spinTime = 0;
+  };
+
+  const handleBlur = (e: React.FocusEvent<HTMLInputElement>) => {
+    commit(e, currentValue, textValue);
+  };
+
+  const handleKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
```

Comment:
Since `commit` happens on blur, I think `Enter` needs to be explicitly handled to also commit (e.g. if someone submits the form using enter from a spinbutton, blur will never happen before the form submits).